src/main/java/org/openjdk/jextract/impl/FunctionalInterfaceBuilder.java

```diff
@@ -80,22 +80,21 @@ static MemorySegment allocate(\{className()} fi, Arena scope) {
             """);
     }
 
-    private void emitFunctionalFactoryForPointer() {
+    private void emitInvoke() {
         boolean needsAllocator = Utils.isStructOrUnion(funcType.returnType());
-        String allocArg = needsAllocator ? ", (SegmentAllocator)arena" : "";
+        String allocParam = needsAllocator ? ", SegmentAllocator alloc" : "";
+        String allocArg = needsAllocator ? ", alloc" : "";
+        String paramStr = methodType.parameterCount() != 0 ? STR.",\{paramExprs()}" : "";
         appendIndentedLines(STR."""
 
             MethodHandle DOWN$MH = Linker.nativeLinker().downcallHandle($DESC);
 
-            static \{className()} ofAddress(MemorySegment addr, Arena arena) {
-                MemorySegment symbol = addr.reinterpret(arena, null);
-                return (\{paramExprs("_")}) -> {
-                    try {
-                        \{retExpr()} DOWN$MH.invokeExact(symbol\{allocArg}\{otherArgExprs()});
-                    } catch (Throwable ex$) {
-                        throw new AssertionError("should not reach here", ex$);
-                    }
-                };
+            static \{methodType.returnType().getSimpleName()} invoke(MemorySegment funcPtr\{allocParam}\{paramStr}) {
```

Comment:
this should be `public` right?

Reply:
Yes. It's implicitly `public` (due to being inside an interface).

Reply:
ah, of course!